examples/full_pipeline.rs

```diff
@@ -111,53 +93,38 @@ impl SumOperator {
     }
 }
 
-struct SumOperatorState {
-    counter: usize,
-    current_timestamp: Timestamp,
-}
-
-#[allow(dead_code)]
-impl SumOperatorState {
-    fn new() -> Self {
-        Self {
-            counter: 0,
-            current_timestamp: Timestamp::Bottom,
-        }
-    }
-
-    fn increment_counter(&mut self, value: usize) {
-        self.counter += value;
-    }
-
-    fn get_counter(&self) -> usize {
-        self.counter
-    }
-}
-
-impl StateT for SumOperatorState {
-    fn commit(&mut self, timestamp: &Timestamp) {
-        self.current_timestamp = timestamp.clone();
-    }
-
-    fn get_last_committed_timestamp(&self) -> Timestamp {
-        self.current_timestamp.clone()
-    }
-}
-
-impl OneInOneOut<SumOperatorState, usize, usize> for SumOperator {
-    fn on_data(&mut self, ctx: &mut OneInOneOutContext<SumOperatorState, usize>, data: &usize) {
+impl OneInOneOut<TimeVersionedState<usize>, usize, usize> for SumOperator {
+    fn on_data(
+        &mut self,
+        ctx: &mut OneInOneOutContext<TimeVersionedState<usize>, usize>,
+        data: &usize,
+    ) {
         tracing::info!("SumOperator @ {:?}: Received {}", ctx.get_timestamp(), data);
 
         let timestamp = ctx.get_timestamp().clone();
-        ctx.get_state().increment_counter(*data);
-        let state = ctx.get_state().get_counter();
+        let timestamp_clone = ctx.get_timestamp().clone();
+
+        // Find the last committed state, add the received data to it, and save for this timestamp.
+        {
+            let past_state = ctx
+                .get_past_state(&ctx.get_last_committed_timestamp())
+                .unwrap();
+            *ctx.get_current_state().unwrap() += past_state + data;
+        }
+
+        // Send the message.
+        let current_state = *ctx.get_current_state().unwrap();
```

Comment:
As-is, this copies a value, so mutating the `current_state` variable will not mutate the operator's state. Keeping `current_state` as a reference, and de-referencing/copying when creating the message should make things more clear.
```suggestion
        let current_state = ctx.get_current_state().unwrap();
```

Reply:
This creates a problem with multiple mutable copies since we use `ctx.get_write_stream` below.

Reply:
Oh, I see. Can we call it `current_state_copy` to clarify?